java/client/src/test/java/io/vitess/client/RpcClientTest.java

```diff
@@ -144,6 +150,45 @@ public void setUp() {
     return values;
   }
 
+  /**
+   * waitForVtgateclienttest waits up to 60 seconds for the "vtgateclienttest" binary to come up
+   * and be reachable on the RPC port.
+   *
+   * We use the "GetSrvKeyspace" RPC exemplary to test if vtgate responds.
+   *
+   * @throws SQLException
+   * @throws InterruptedException
+   */
+  private void waitForVtgateclienttest() throws SQLException, InterruptedException {
+    DateTime start = DateTime.now();
+    DateTime deadline = start.plusSeconds(60);
+
+    boolean waited = false;
+    while (DateTime.now().isBefore(deadline)) {
+      try {
+        ctx = Context.getDefault().withDeadlineAfter(Duration.millis(5000)).withCallerId(CALLER_ID);
+        conn.getSrvKeyspace(ctx, "small");
+        // RPC succeeded. Stop testing.
+        break;
+      } catch (SQLTransientException e) {
+        Throwable rootCause = Throwables.getRootCause(e);
+        if (!rootCause.getMessage().contains("Connection refused: ")) {
+          // Non-retryable exception. Fail for good.
+          throw e;
+        }
+
+        System.out.format("Waiting until vtgateclienttest is ready and responds (got exception: %s)\n", rootCause);
+        Thread.sleep(TimeUnit.MILLISECONDS.toMillis(100));
```

Comment:
A note here:

Nothing in this method influences the RPC "connection". Instead, the context above sets a deadline on the RPC call itself and the RPC subsystem should enforce that i.e. return after the deadline has been exceeded.

In general, RPCs are agnostic of any underlying connection: They only guarantee you to send out a request and return a response. They don't care how this is actually implemented.

For example, the RPC subsystem could close idle TCP connections and automatically reestablish them on demand.

I've changed the javadoc comment of the method to better explain that we're constantly trying to execute the RPC until it finally succeeded. I hope this made the approach clearer :)